### ISSUE TITLE
Templating: Set variable values via url and have them stay even when it's not an available option given the template variable query results 

### ISSUE THREAD

**Issue by aurimasplu** [grafana/grafana#21895]
**What happened**:
Upgraded Grafana from v6.1.6 to v6.6.0.
After upgrade, when clicking on link in one dashboard, new dashboard opens and loads,variables seems to be included correctly, but but after one or two seconds dashboard reloads to its default variables to which it was last saved.

**What you expected to happen**:
Dashboard loads and stays with inherited variables. 

**How to reproduce it (as minimally and precisely as possible)**:
Create two dashboards which shares same variables and add link with tags from one dashboard to another. 
**Anything else we need to know?**:

**Environment**:
- Grafana version: 6.6.0
- Data source type & version: Elasticsearch 6.6
- OS Grafana is installed on: RHEL 7.7
- User OS & Browser: Win 10, Chrome, Firefox
- Grafana plugins: Default panels
- Others:

**Comment by peterholmberg** [comment]
Thank you for reporting this @aurimasplu. Could you provide any errors or warnings from the developer tools console in your browser?

**Comment by dprokop** [comment]
@aurimasplu thanks for reporting. Just tried to reproduce this issue, but seems to be working fine in 6.6.0. Do you mind sharing some more details about your dashboard?

**Comment by aurimasplu** [comment]
Nothing spatial in console:
![image](https://user-images.githubusercontent.com/33485896/73767604-570bd280-4780-11ea-9b86-8b3b09200a3e.png)

**Comment by aurimasplu** [comment]
This is how it looks:
![image](https://user-images.githubusercontent.com/33485896/73767950-dd281900-4780-11ea-8392-e579d110c96a.png)

**Comment by tskarhed** [comment]
I cannot reproduce. Tried going to `v6.1.x` and created dashboards recreating the scenario, and switched to `v6.6.x`. Does the variable have the same value options on both dashboards? Is it possible you could share a JSON of the dashboards, alternatively reproduce an minimal case?

Thank you for the screenshots, it was very helpful! :)

**Comment by aurimasplu** [comment]
I have created two complacently fresh dashboards and upgrades to Grafana 6.6.1. Same behavior.


Please see JSON's below:

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 297,
  "iteration": 1581338289656,
  "links": [
    {
      "icon": "external link",
      "includeVars": true,
      "keepTime": true,
      "tags": [
        "testInterface"
      ],
      "targetBlank": true,
      "type": "dashboards"
    }
  ],
  "panels": [
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "Elastic-Telegraf-System",
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "avg": false,
        "current": false,
        "hideEmpty": true,
        "hideZero": true,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "dataLinks": []
      },
      "percentage": false,
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "bucketAggs": [
            {
              "field": "@timestamp",
              "id": "2",
              "settings": {
                "interval": "auto",
                "min_doc_count": 0,
                "trimEdges": 0
              },
              "type": "date_histogram"
            }
          ],
          "metrics": [
            {
              "field": "CPU.cpmCPUTotal1min",
              "id": "1",
              "meta": {},
              "settings": {},
              "type": "avg"
            }
          ],
          "query": "tag.agent_host:\"$Device\"",
          "refId": "A",
          "timeField": "@timestamp"
        },
        {
          "bucketAggs": [
            {
              "field": "@timestamp",
              "id": "2",
              "settings": {
                "interval": "auto",
                "min_doc_count": 0,
                "trimEdges": 0
              },
              "type": "date_histogram"
            }
          ],
          "metrics": [
            {
              "field": "CPU.cpmCPUTotal1minRev",
              "id": "1",
              "meta": {},
              "settings": {},
              "type": "avg"
            }
          ],
          "query": "tag.agent_host:\"$Device\"",
          "refId": "B",
          "timeField": "@timestamp"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "CPU",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "decimals": null,
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": "100",
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    }
  ],
  "schemaVersion": 22,
  "style": "dark",
  "tags": [
    "testDevice"
  ],
  "templating": {
    "list": [
      {
        "allValue": "*",
        "current": {
          "selected": true,
          "text": "All",
          "value": "$__all"
        },
        "datasource": "Elastic-Telegraf-System",
        "definition": "{\"find\": \"terms\", \"field\": \"tag.platform_tag\",\"size\":6000}",
        "hide": 0,
        "includeAll": true,
        "label": "Platform",
        "multi": false,
        "name": "Platform",
        "options": [],
        "query": "{\"find\": \"terms\", \"field\": \"tag.platform_tag\",\"size\":6000}",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 1,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      },
      {
        "allValue": null,
        "current": {
          "selected": true,
          "text": "xxxxx.mynet.net",
          "value": "xxxxx.mynet.net"
        },
        "datasource": "Elastic-Telegraf-System",
        "definition": "{\"query\": \"tag.platform_tag:$Platform\",\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}",
        "hide": 0,
        "includeAll": false,
        "label": "Device",
        "multi": false,
        "name": "Device",
        "options": [],
        "query": "{\"query\": \"tag.platform_tag:$Platform\",\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 1,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ]
  },
  "timezone": "",
  "title": "Test Device",
  "uid": "YeBiQQ8Zk",
  "version": 7
}
```


   





```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 298,
  "iteration": 1581338435956,
  "links": [],
  "panels": [
    {
      "aliasColors": {},
      "bars": false,
      "dashLength": 10,
      "dashes": false,
      "datasource": "Elastic-Telegraf-Interface",
      "fill": 1,
      "fillGradient": 0,
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "hiddenSeries": false,
      "id": 2,
      "legend": {
        "avg": false,
        "current": false,
        "max": false,
        "min": false,
        "show": true,
        "total": false,
        "values": false
      },
      "lines": true,
      "linewidth": 1,
      "nullPointMode": "null",
      "options": {
        "dataLinks": []
      },
      "percentage": false,
      "pointradius": 2,
      "points": false,
      "renderer": "flot",
      "seriesOverrides": [],
      "spaceLength": 10,
      "stack": false,
      "steppedLine": false,
      "targets": [
        {
          "alias": "IN Bits",
          "bucketAggs": [
            {
              "field": "@timestamp",
              "id": "2",
              "settings": {
                "interval": "1m",
                "min_doc_count": 0,
                "trimEdges": 0
              },
              "type": "date_histogram"
            }
          ],
          "metrics": [
            {
              "field": "interface.ifHCInOctets",
              "hide": true,
              "id": "1",
              "inlineScript": "_value*8",
              "meta": {},
              "settings": {
                "script": {
                  "inline": "_value*8"
                }
              },
              "type": "max"
            },
            {
              "field": "1",
              "id": "3",
              "meta": {},
              "pipelineAgg": "1",
              "settings": {
                "unit": "1s"
              },
              "type": "derivative"
            }
          ],
          "query": "tag.agent_host:\"$Device\" AND tag.ifDescr:\"$Interface\"",
          "refId": "A",
          "timeField": "@timestamp"
        }
      ],
      "thresholds": [],
      "timeFrom": null,
      "timeRegions": [],
      "timeShift": null,
      "title": "Interface Utilization",
      "tooltip": {
        "shared": true,
        "sort": 0,
        "value_type": "individual"
      },
      "type": "graph",
      "xaxis": {
        "buckets": null,
        "mode": "time",
        "name": null,
        "show": true,
        "values": []
      },
      "yaxes": [
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        },
        {
          "format": "short",
          "label": null,
          "logBase": 1,
          "max": null,
          "min": null,
          "show": true
        }
      ],
      "yaxis": {
        "align": false,
        "alignLevel": null
      }
    }
  ],
  "schemaVersion": 22,
  "style": "dark",
  "tags": [
    "testInterface"
  ],
  "templating": {
    "list": [
      {
        "allValue": null,
        "current": {
          "selected": false,
          "text": "yyyyyyy.mynet.net",
          "value": "yyyyyyy.mynet.net"
        },
        "datasource": "Elastic-Telegraf-Interface",
        "definition": "{\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}",
        "hide": 0,
        "includeAll": false,
        "label": "Device",
        "multi": false,
        "name": "Device",
        "options": [],
        "query": "{\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 1,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      },
      {
        "allValue": null,
        "current": {
          "text": "Ethernet1/1",
          "value": "Ethernet1/1"
        },
        "datasource": "Elastic-Telegraf-Interface",
        "definition": "{\"find\": \"terms\", \"field\": \"tag.ifDescr\", \"query\": \"tag.agent_host:$Device\",\"size\": 6000}",
        "hide": 0,
        "includeAll": false,
        "label": "Interface",
        "multi": false,
        "name": "Interface",
        "options": [],
        "query": "{\"find\": \"terms\", \"field\": \"tag.ifDescr\", \"query\": \"tag.agent_host:$Device\",\"size\": 6000}",
        "refresh": 1,
        "regex": "",
        "skipUrlSync": false,
        "sort": 0,
        "tagValuesQuery": "",
        "tags": [],
        "tagsQuery": "",
        "type": "query",
        "useTags": false
      }
    ]
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ]
  },
  "timezone": "",
  "title": "Test Interface",
  "uid": "E7ypwQUWk",
  "version": 8
}
```

**Comment by AjitVV** [comment]
Hi All,

We had Grafana v 6.2.0 before. After upgrading to 6.6.0, the navigation between dashboards seems to be not working properly, that is, while going from one dashboard to next one, the variable drop down selection seems to be reset to default. We observed that the URL changes automatically, as a result the variables get reset. This issue is not observed always, it happens randomly after few navigation.

We are also facing the same issue in v 6.6.1 with the new and the existing dashboards. Is there a planned fixed for this issue?

**Comment by tskarhed** [comment]
Issue successfully reproduced with TestData in master.

1.  The same name for two "different" query variables on two differrent dashboards
2. Their queries differ
3. They have refresh set to on dashboard load
4. A dashboard link with tag between the two is set to pass on template variables

**What happens:**
The value is successfully passed on, but is replaced when the query for the query variable (with a different query) has returned.

**Solution suggestion:**
Give template vars in URL higher priority

**Comment by tskarhed** [comment]
> Hi All,
> 
> We had Grafana v 6.2.0 before. After upgrading to 6.6.0, the navigation between dashboards seems to be not working properly, that is, while going from one dashboard to next one, the variable drop down selection seems to be reset to default. We observed that the URL changes automatically, as a result the variables get reset. This issue is not observed always, it happens randomly after few navigation.
> 
> We are also facing the same issue in v 6.6.1 with the new and the existing dashboards. Is there a planned fix for this issue?

Are you using query variables and if you do, what is your refresh set to?

**Comment by AjitVV** [comment]
> > Hi All,
> > We had Grafana v 6.2.0 before. After upgrading to 6.6.0, the navigation between dashboards seems to be not working properly, that is, while going from one dashboard to next one, the variable drop down selection seems to be reset to default. We observed that the URL changes automatically, as a result the variables get reset. This issue is not observed always, it happens randomly after few navigation.
> > We are also facing the same issue in v 6.6.1 with the new and the existing dashboards. Is there a planned fix for this issue?
> 
> Are you using query variables and if you do, what is your refresh set to?

Yes, we are using query variables and refresh is set to "On Time Range Change"

**Comment by praveenacharyab** [comment]
We are also facing this issue. We had 6.5.0 version and upgraded to 6.6.1 and the navigation links seems to be broken. When I click the links from table panel it goes to the target dashboard. But even though variables and values are passed as query params in url, the target dashboard loads but quickly resets to default variables (dropdown at top left corner) and also the url in browser address bar changes. We are facing this issue when clicking on any cell value rendered as url in table panel (click on on any cell value in the table panel), it navigates to next dashboard but variables are immediately reset.

**Comment by tskarhed** [comment]
The functionality related to this is implemented here.:
https://github.com/grafana/grafana/blob/fa1602daea7dbe3d96bb9bfde9634f8e5ede8591/public/app/features/templating/variable_srv.ts#L270

Still investigating.

**Comment by Jessimon** [comment]
I also have this issue. I upgraded from version 6.5.1 to 6.6.1.
I did notice 3 things:
- I only have the issue on my Live Grafana with dashboards that i created on my test Grafana server. On these dashboards the navigation resets to a $host that is not even in the Prometeus database.
- When i choose a $host on a dashboard and then navigate to a dashboard that has the navigation issue the initial $host is shown en then it switches to the host that does not exists.
- If i navigate between dashboards that have been created on the Live server there is no problem.

This is the variable that i use:
label_values({job=~"$job"}, instance)
It queries a Prometheus datasource.

**Comment by tskarhed** [comment]
@Jessimon Do you have the same query variable for all dashboard (with the same query)?

**Comment by Jessimon** [comment]
@tskarhed Yes, all the dashboards use the same query and variable settings.

![image](https://user-images.githubusercontent.com/61043214/74647056-45143180-517b-11ea-8d52-be96b2d210f2.png)

**Comment by torkelo** [comment]
It sounds like the option being set from URL is not returned by the query on the second dashboard, so grafana detects that it's not a valid value and set's the value to the default (first) returned value from the query. 

The Device variable query is different in the two queries: 
First dashboard: 
```{\"query\": \"tag.platform_tag:$Platform\",\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}```
Second dashboard: 
```{\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}```

So if the second dashboard has a different device query it will update, and trigger an update in the dependent variable (interface) and if the value set from url is not available in the result it will be rest.

**Comment by tskarhed** [comment]
Yes, but then comes a few questions:
 
1. Should the value in the URL take precedence over whatever default value is set, regardless of whether or not the query returns it? I say yes - this issue is a proof there is a use case for it
2. Why did this work before?

Also, if you look at @Jessimon issue, they do have the same query.

**Comment by Jessimon** [comment]
I updates to version 6.6.2, but that does not change anything.

**Comment by tskarhed** [comment]
@torkelo What's your decision on this? I really feel that this needs fixing.

**Comment by torkelo** [comment]
Do you have two dashboards with the same template variable query and links between where you can replicate this?

**Comment by tskarhed** [comment]
No, I can't manage to replicate when the query is the same. It's rather about when you have two different queries that the URL (when using a dashboard link, if that's applicable) overrides whatever the other query returns. It might be more of a feature request, but since the functionality has existed before, it's definitely a regression.

**Comment by torkelo** [comment]
yea, if you have two different queries, and the second query does not return what your trying to set via url then that is not a bug but how the current templating system works.

**Comment by Jessimon** [comment]
> Do you have two dashboards with the same template variable query and links between where you can replicate this?

All my dashboards use the same template variable query and are all linked.
If that can help we can setup some Hangout so that i can show it to you.

At the moment Grafana is very confusing to use because people do not always notice that the dashboard they navigated to is not showing the data from the server they expect because the server has been reset by the navigation.

**Comment by torkelo** [comment]
Can you use chrome dev tools and inspect the network queries that are coming from your query variables? What are their responses? Do the response contain the value set from url?

**Comment by Jessimon** [comment]
Never used that tool.
What do i need to do?

**Comment by Jessimon** [comment]
Would it help if i give you access to our Grafana?
Then you can see whats happening.

**Comment by Jessimon** [comment]
I was able to partially solve the problem.
I deleted all my variables and recreated them.
On dashboards where i don't need the $host variable i change the variable from
label_values({job=~"$job"}, instance)
to
label_values(instance)
On these dashboards there are no issues any more. The choosen host stays the same when navigation from one dashboard to an other.
On dashboards that do need a filter on that variable, i changed the variable from:
label_values({job=~"$job"}, instance)
to
label_values({job="netdata"}, instance)
That works to as long as i don't select refresh: On time range change.

If i use refresh: On time range change one of two things happen:
The dashboard takes a very long time to load and then i get the message "Template variable service failed".
The dashboard goes haywire and the data on the dashboard is constantly changing wildly.

Looks to me like there is an issue with variables.

The variables that i use are related to Prometheus.
I am currently using version 2.14.

**Comment by dp106** [comment]
Issue is still present in 6.7.1

**Comment by aurimasplu** [comment]
> It sounds like the option being set from URL is not returned by the query on the second dashboard, so grafana detects that it's not a valid value and set's the value to the default (first) returned value from the query.
> 
> The Device variable query is different in the two queries:
> First dashboard:
> `{\"query\": \"tag.platform_tag:$Platform\",\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}`
> Second dashboard:
> `{\"find\": \"terms\", \"field\": \"tag.agent_host\",\"size\": 6000}`
> 
> So if the second dashboard has a different device query it will update, and trigger an update in the dependent variable (interface) and if the value set from url is not available in the result it will be rest.

In my case both variable queries return exactly the same results, second one could be considered wider as it is not filtered by Platform. In both cases same data source is used.
Also if you look at my earlier provided screenshots, you would see that correct data is loaded at first. But later refreshes to default.

**Comment by dp106** [comment]
Has anyone tested this in version 7?  Not able to reproduce it... Yet

**Comment by tskarhed** [comment]
Seems to be fixed in master and v7.0.x branch, so I'll close this issue.
